scripts/bluetooth/mesh/mesh_dfu_metadata.py

```diff
@@ -0,0 +1,425 @@
+            comp (dict): A Composition Data dictionary class object to be filled in
+
+        Returns:
+            TBA
+        """
+        label_cnt = int(kconfigs['CONFIG_BT_MESH_LABEL_COUNT']) if 'CONFIG_BT_MESH_LABEL_COUNT' in kconfigs.keys() else 0
+        lcd_srv = (kconfigs['CONFIG_BT_MESH_LARGE_COMP_DATA_SRV'] == 'y') if 'CONFIG_BT_MESH_LARGE_COMP_DATA_SRV' in kconfigs.keys() else False
+
+        with open(self.elf_path, 'rb') as elf_file:
+            elf = ELFFile(elf_file)
+            cdp0_value = self.__read_symbol_data(elf, addr)
+
+            # The format of the composition data is defined by `struct bt_mesh_comp` type.
+            # The format of an element is defined by `struct bt_mesh_elem` type.
+            # The format of a model is defined by `struct bt_mesh_model` type.
+            # All types are declared in `zephyr/include/zephyr/bluetooth/mesh/access.h`.
+            cid, pid, vid, _, elems_count, elems_ptr = struct.unpack('HHHHII', cdp0_value)
+
+            comp.cid = cid
+            comp.pid = pid
+            comp.vid = vid
+
+            elems_value = self.__read_symbol_data(elf, elems_ptr)
+            elems_iter = struct.iter_unpack('IHBBII', elems_value)
+            i = 0
+
+            for elem in elems_iter:
+                i += 1
+                if i > elems_count:
+                    print("Extracted more elems than \'elem_count\'")
+                    sys.exit(1)
+                __rt, loc, sig_count, vnd_count, sig_ptr, vnd_ptr = elem
+                elem_item = comp.elem_add(loc)
+
+                def models_unpack(ptr, elem_item, vnd):
+                    models_value = self.__read_symbol_data(elf, ptr)
+                    model_format = 'HHIIIHHIHH' + ('I' if label_cnt > 0  else '') + 'II' + ('I' if lcd_srv else '')
+                    models_iter = struct.iter_unpack(model_format, models_value)
+
+                    for model in models_iter:
+                        id1, id2, __rt, __pub, __keys, __keys_cnt, _, __groups, __groups_cnt, _, __uuids, __op, __cb = model
+                        if not vnd:
+                            elem_item.sig_model_add(id1)
+                        else:
+                            elem_item.vnd_model_add(id1, id2)
+
+                if sig_count > 0:
+                    models_unpack(sig_ptr, elem_item, False)
+                if vnd_count > 0:
+                    models_unpack(vnd_ptr, elem_item, True)
+
+    def extract_comp_data(self, kconfigs, comp):
+        """
+        Extract composition data from .elf file.
+
+        Parameters:
+            kconfigs (dict): A dictionary representing Kconfig options used for the firmware to compile with
+            comp (dict): A Composition Data dictionary class object to be filled in
+
+        Returns:
+            TBA
+        """
+        for addr in self.addrs:
+            self.__read_comp_data(addr, kconfigs, comp)
+
```

Comment:
I might be misreading, but `__read_comp_data` will _overwrite_ `cid`, `pid` and `vid`, and then just append the elements to what is already there in `comp`. Meaning if there are multiple `addrs`, `comp` will contain the `cid` etc.from the last one parsed, and all the elements from all the composition data instances?

In general it is a bit weird for this function to take in `comp` at all, looks like writing C in Python. I suggest to rewrite `__read_comp_data` to return a `Comp0` directly, and do

```python
return [self.__read_comp_data(addr, kconfigs) for addr in self.addrs]
```

or something here. Then the caller will have to handle potentially multiple instances.

Reply:
Yes... I think you are right. I had a feeling that something was a bit off.